File: backend/app/schemas/graph_spec.py
from pydantic import BaseModel, Field
from typing import List, Optional, Union, Dict, Any
from enum import Enum

class GraphType(str, Enum):
    TWO_D_FUNCTION = "2d_function"
    TWO_D_PARAMETRIC = "2d_parametric"
    TWO_D_IMPLICIT = "2d_implicit"
    THREE_D_SURFACE = "3d_surface"
    THREE_D_PARAMETRIC = "3d_parametric"
    POINTS_ONLY = "points_only"

class TraceKind(str, Enum):
    SCATTER = "scatter"
    SCATTER_3D = "scatter3d"
    SURFACE = "surface"
    CONTOUR = "contour"

class GraphTrace(BaseModel):
    name: str
    kind: TraceKind
    x: List[Optional[float]]
    y: List[Optional[float]]
    z: Optional[List[Optional[float]]] = None
    # For surface/contour, z is often a 2D matrix
    z_matrix: Optional[List[List[Optional[float]]]] = None
    mode: str = "lines" # lines, markers, lines+markers
    line_style: Optional[Dict[str, Any]] = None
    marker_style: Optional[Dict[str, Any]] = None
    show_legend: bool = True

class KeyPoint(BaseModel):
    label: str
    x: float
    y: float
    z: Optional[float] = None
    color: Optional[str] = None

class GraphAxes(BaseModel):
    x_label: str = "x"
    y_label: str = "y"
    z_label: Optional[str] = None
    x_range: Optional[List[float]] = None
    y_range: Optional[List[float]] = None
    z_range: Optional[List[float]] = None
    equal_aspect: bool = False

class GraphSpec(BaseModel):
    version: str = "1.0"
    graph_type: GraphType
    title: Optional[str] = None
    axes: GraphAxes = Field(default_factory=GraphAxes)
    traces: List[GraphTrace] = Field(default_factory=list)
    key_points: List[KeyPoint] = Field(default_factory=list)
    sampling_info: Optional[Dict[str, Any]] = None
    warnings: List[str] = Field(default_factory=list)

File: src/components/workspace/graph_spec.ts
export type GraphType =
    | "2d_function"
    | "2d_parametric"
    | "2d_implicit"
    | "3d_surface"
    | "3d_parametric"
    | "points_only";

export type TraceKind = "scatter" | "scatter3d" | "surface" | "contour";

export interface GraphTrace {
    name: string;
    kind: TraceKind;
    x: (number | null)[];
    y: (number | null)[];
    z?: (number | null)[];
    z_matrix?: (number | null)[][];
    mode?: "lines" | "markers" | "lines+markers";
    line_style?: Record<string, unknown>;
    marker_style?: Record<string, unknown>;
    show_legend?: boolean;
}

export interface KeyPoint {
    label: string;
    x: number;
    y: number;
    z?: number;
    color?: string;
}

export interface GraphAxes {
    x_label: string;
    y_label: string;
    z_label?: string;
    x_range?: [number, number];
    y_range?: [number, number];
    z_range?: [number, number];
    equal_aspect?: boolean;
}

export interface GraphSpec {
    version: string;
    graph_type: GraphType;
    title?: string;
    axes: GraphAxes;
    traces: GraphTrace[];
    key_points: KeyPoint[];
    sampling_info?: Record<string, unknown>;
    warnings: string[];
}
